solcore/optics/beer_lambert.py

```diff
@@ -106,16 +110,24 @@ def alf(x):
 
     # Each building block (layer or junction) needs to have access to the absorbed light in its region.
     # We update each object with that information.
-    I_0 = fraction
+
+    I0 = 1*fraction # need the *1 because we DO NOT want to modify fraction! Will mess up profile calculation
+
+    layers_above_offset = np.cumsum([0] + n_layers_junction)
+
     for j in range(len(solar_cell)):
         solar_cell[j].diff_absorption = diff_absorption
         solar_cell[j].absorbed = types.MethodType(absorbed, solar_cell[j])
 
         # total absorption at each wavelength, per layer
-        layer_positions = options.position[(options.position >= solar_cell[j].offset) & (
-                options.position < solar_cell[j].offset + solar_cell[j].width)]
-        layer_positions = layer_positions - np.min(layer_positions)
-        solar_cell[j].layer_absorption = np.trapz(solar_cell[j].absorbed(layer_positions), layer_positions, axis=0)
+        A_junc = np.zeros_like(wl_m)
+
+        for k in range(n_layers_junction[j]):
```

Comment:
This is very well thought! Getting rid of `trapz` will save us from numerical errors and possibly even speed things up. 

If anything, for clarity, I would define a variable `ilayer = layers_above_offset[j] + k` and use this in the formulas, so they are less verbose.